multinode cluster: fix waits and joins

prezha opened the pull request:
fixes #10640

multinode tests usually fail for at least these three reasons:
1. we wait with WaitForNodeReady in WaitForNode, which essentially waits for all nodes instead of just the one that is created, which could be the problem after the multinode cluster is restarted; it timeouts after 6mins with 'node has unwanted condition "Ready"', waiting for an 'old' node to rejoin - a problem described in (3.) below
2. we wait for all system pods, and that could also timeout after 6mins - in case of the multinode cluster restart, waiting sometimes also for those stale leftovers (ie, pods that are 'remembered' on nodes that were removed or not yet restarted)
3. we try to 'kubeadm reset' then 'kubeadm join' existing nodes, and this is explicitly 'forbidden':
- 'reject a node joining the cluster if a node with the same name already exists (1711)': https://github.com/kubernetes/kubeadm/issues/1711
("kubeadm should take action on preventing the bootstrap of such nodes on kubeadm join time.")
- 'kubeadm: prevent bootstrap of nodes with known names (81056)'
https://github.com/kubernetes/kubernetes/pull/81056
("If a Node name in the cluster is already taken and this Node is Ready, prevent TLS bootsrap on "kubeadm join" and exit early.")

so, trying to rejoin 'old' node timeouts after 3mins with 'error execution phase kubelet-start: a Node with name "_name_" and status "Ready" already exists in the cluster. You must delete the existing Node or change the name of this new joining Node', then deleting that node and creating a new one that can then join the cluster w/o issues (but it all takes additional time)

now, this pr addresses these issues by:
1. waiting only for a node that has been currently worked on
2. restructured 'waiting for pods' logic by also checking some additional indicators to avoid api responses with stale data, while reducing the total number of api calls
3. restructured 'join cluster' logic: removed 'reset' part and added 'remove' operation that drains then removes 'old' node from the cluster before trying to add it back + some additional flags to prevent ending up stuck in the process (eg, 'error: unable to drain node "multinode-777-m03", aborting command... There are pending nodes to be drained: error: cannot delete DaemonSet-managed Pods (use --ignore-daemonsets to ignore)')

i've successfully tested multinode and functional suites with docker and kvm drivers, and, as we avoid unnecessary waits and timeouts, they should now also be significantly faster, but let's wait for official results from @medyagh  ;)

prezha commented on pkg/provision/ubuntu.go:
```diff
@@ -84,7 +84,7 @@ StartLimitIntervalSec=60
 [Service]
 Type=notify
-Restart=on-failure
+Restart=always
```
yes, you are right, i'll remove it from this pr

prezha replied:
removed

prezha commented on pkg/minikube/bootstrapper/kubeadm/kubeadm.go:
```diff
@@ -485,8 +486,18 @@ func (k *Bootstrapper) WaitForNode(cfg config.ClusterConfig, n config.Node, time
 		return nil
 	}
 
+	if cfg.VerifyComponents[kverify.NodeReadyKey] {
```
this node ready check was just moved further ahead within the same WaitForNode func, and as we are already checking it, made more sense to do it before we check for the health of the pods of that node;

an additional difference is that this new one only checks for the readiness of that specific node, and the previous one checked all the nodes (and that created some of the issues with restarted multinode cluster

it was originally added with pr #7752

prezha commented on pkg/minikube/node/node.go:
```diff
@@ -78,8 +78,8 @@ func Add(cc *config.ClusterConfig, n config.Node, delOnFail bool) error {
-// Delete stops and deletes the given node from the given cluster
-func Delete(cc config.ClusterConfig, name string) (*config.Node, error) {
+// remove drains and deletes the given node in the given cluster.
+func remove(cc config.ClusterConfig, name string) (*config.Node, error) {
```
in addition to draining a node first, we are also removing it from the cluster right after - should we name the func eg removeNodeFromCluster?

medyagh replied:
I think drainNode is a better name. since it is called by the Delete func is okay, I just want more specific names, so I know this func does drain and this other one actually deletes. so lets go with drainNode

prezha replied:
sure, i've renamed it

prezha commented on pkg/minikube/node/node.go:
```diff
@@ -122,13 +124,29 @@ func Delete(cc config.ClusterConfig, name string) (*config.Node, error) {
+	m := config.MachineName(cc, *n)
+	api, err := machine.NewAPIClient()
+	if err != nil {
+		return n, err
```
maybe i'm wrong, but in this place, it will return if it cannot get api client (needed for delete), not if it cannot load the config?

did you perhaps referred to the previous call to remove func? i do agree it should not return on err there, but continue and try to delete the host anyway, just not sure atm if we would also need to do anything additionally to clean up the leftovers

medyagh commented on pkg/minikube/node/node.go:
```diff
@@ -122,13 +124,29 @@ func Delete(cc config.ClusterConfig, name string) (*config.Node, error) {
 
-	cc.Nodes = append(cc.Nodes[:index], cc.Nodes[index+1:]...)
-	return n, config.SaveProfile(viper.GetString(config.ProfileName), &cc)
+	return n, nil
```
what was the reason that was removed from this PR ? was it related ? or could it be done in a separate PR ?

medyagh commented:
/ok-to-test

medyagh commented:
@prezha the none driver test says this which seems odd, since in NoneDriver there is no Multi-node 
```
	X Exiting due to GUEST_START: wait 6m0s for node: waiting for node to be ready: waitNodeCondition: error getting node "minikube-m01": nodes "minikube-m01" not found
```

prezha commented:
> @prezha the none driver test says this which seems odd, since in NoneDriver there is no Multi-node
> none_Linux — Jenkins: completed with 10 / 83 failures in 23.01 minute(s). https://storage.googleapis.com/minikube-builds/logs/10758/622b4e3/none_Linux.html
> 
> ```
> 	X Exiting due to GUEST_START: wait 6m0s for node: waiting for node to be ready: waitNodeCondition: error getting node "minikube-m01": nodes "minikube-m01" not found
> ```
> 
> might also wanna rebase from upstream

good spot, thank you @medyagh !

here is what i've found:

- [in **sigle-node clusters**] with `none` driver, we have ***node name set to "m01"*** - this example:
```
2021-03-10T01:15:04.7548443Z         	* I0310 01:14:00.604410    5764 kubeadm.go:386] StartCluster: {Name:minikube KeepContext:false EmbedCerts:false MinikubeISO: KicBaseImage:gcr.io/k8s-minikube/kicbase:v0.0.18@sha256:ddd0c02d289e3a6fb4bba9a94435840666f4eb81484ff3e707b69c1c484aa45e Memory:4000 CPUs:2 DiskSize:20000 VMDriver: Driver:none HyperkitVpnKitSock: HyperkitVSockPorts:[] DockerEnv:[] ContainerVolumeMounts:[] InsecureRegistry:[] RegistryMirror:[] HostOnlyCIDR:192.168.99.1/24 HypervVirtualSwitch: HypervUseExternalSwitch:false HypervExternalAdapter: KVMNetwork:default KVMQemuURI:qemu:///system KVMGPU:false KVMHidden:false KVMNUMACount:1 DockerOpt:[] DisableDriverMounts:false NFSShare:[] NFSSharesRoot:/nfsshares UUID: NoVTXCheck:false DNSProxy:false HostDNSResolver:true HostOnlyNicType:virtio NatNicType:virtio SSHIPAddress: SSHUser:root SSHKey: SSHPort:22 KubernetesConfig:{KubernetesVersion:v1.20.2 ClusterName:minikube Namespace:default APIServerName:minikubeCA APIServerNames:[] APIServerIPs:[] DNSDomain:cluster.local ContainerRuntime:docker CRISocket: NetworkPlugin: FeatureGates: ServiceCIDR:10.96.0.0/12 ImageRepository: LoadBalancerStartIP: LoadBalancerEndIP: CustomIngressCert: ExtraOptions:[{Component:kubelet Key:resolv-conf Value:/run/systemd/resolve/resolv.conf}] ShouldLoadCachedImages:false EnableDefaultCNI:false CNI: NodeIP: NodePort:8441 NodeName:} Nodes:[{Name:m01 IP:10.1.0.4 Port:8441 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}] Addons:map[] VerifyComponents:map[apiserver:true apps_running:true default_sa:true extra:true kubelet:true node_ready:true system_pods:true] StartHostTimeout:6m0s ScheduledStop:<nil> ExposedPorts:[] ListenAddress: Network: MultiNodeRequested:false}
```
extract:
> Nodes:[{Name:m01 IP:10.1.0.4 Port:8441 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}]

in contrast, in **sigle-node clusters** with `kvm2` and `docker` drivers, we have ***node name unset***:

- kvm2 driver - example:
```
I0310 15:36:39.599866   26531 kubeadm.go:386] StartCluster: {Name:minikube KeepContext:false EmbedCerts:false MinikubeISO:https://storage.googleapis.com/minikube/iso/minikube-v1.18.0.iso KicBaseImage:gcr.io/k8s-minikube/kicbase:v0.0.18@sha256:ddd0c02d289e3a6fb4bba9a94435840666f4eb81484ff3e707b69c1c484aa45e Memory:6000 CPUs:2 DiskSize:20000 VMDriver: Driver:kvm2 HyperkitVpnKitSock: HyperkitVSockPorts:[] DockerEnv:[] ContainerVolumeMounts:[] InsecureRegistry:[] RegistryMirror:[] HostOnlyCIDR:192.168.99.1/24 HypervVirtualSwitch: HypervUseExternalSwitch:false HypervExternalAdapter: KVMNetwork:default KVMQemuURI:qemu:///system KVMGPU:false KVMHidden:false KVMNUMACount:1 DockerOpt:[] DisableDriverMounts:false NFSShare:[] NFSSharesRoot:/nfsshares UUID: NoVTXCheck:false DNSProxy:false HostDNSResolver:true HostOnlyNicType:virtio NatNicType:virtio SSHIPAddress: SSHUser:root SSHKey: SSHPort:22 KubernetesConfig:{KubernetesVersion:v1.20.2 ClusterName:minikube Namespace:default APIServerName:minikubeCA APIServerNames:[] APIServerIPs:[] DNSDomain:cluster.local ContainerRuntime:docker CRISocket: NetworkPlugin: FeatureGates: ServiceCIDR:10.96.0.0/12 ImageRepository: LoadBalancerStartIP: LoadBalancerEndIP: CustomIngressCert: ExtraOptions:[] ShouldLoadCachedImages:true EnableDefaultCNI:false CNI: NodeIP: NodePort:8443 NodeName:} Nodes:[{Name: IP:192.168.39.57 Port:8443 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}] Addons:map[] VerifyComponents:map[apiserver:true apps_running:true default_sa:true extra:true kubelet:true node_ready:true system_pods:true] StartHostTimeout:6m0s ScheduledStop:<nil> ExposedPorts:[] ListenAddress: Network: MultiNodeRequested:false}
```
extract:
> Nodes:[{Name: IP:192.168.39.57 Port:8443 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}]

- docker driver - example:
```
I0310 14:57:22.024808    7610 kubeadm.go:386] StartCluster: {Name:minikube KeepContext:false EmbedCerts:false MinikubeISO: KicBaseImage:gcr.io/k8s-minikube/kicbase:v0.0.18@sha256:ddd0c02d289e3a6fb4bba9a94435840666f4eb81484ff3e707b69c1c484aa45e Memory:16000 CPUs:2 DiskSize:20000 VMDriver: Driver:docker HyperkitVpnKitSock: HyperkitVSockPorts:[] DockerEnv:[] ContainerVolumeMounts:[] InsecureRegistry:[] RegistryMirror:[] HostOnlyCIDR:192.168.99.1/24 HypervVirtualSwitch: HypervUseExternalSwitch:false HypervExternalAdapter: KVMNetwork:default KVMQemuURI:qemu:///system KVMGPU:false KVMHidden:false KVMNUMACount:1 DockerOpt:[] DisableDriverMounts:false NFSShare:[] NFSSharesRoot:/nfsshares UUID: NoVTXCheck:false DNSProxy:false HostDNSResolver:true HostOnlyNicType:virtio NatNicType:virtio SSHIPAddress: SSHUser:root SSHKey: SSHPort:22 KubernetesConfig:{KubernetesVersion:v1.20.2 ClusterName:minikube Namespace:default APIServerName:minikubeCA APIServerNames:[] APIServerIPs:[] DNSDomain:cluster.local ContainerRuntime:docker CRISocket: NetworkPlugin: FeatureGates: ServiceCIDR:10.96.0.0/12 ImageRepository: LoadBalancerStartIP: LoadBalancerEndIP: CustomIngressCert: ExtraOptions:[] ShouldLoadCachedImages:true EnableDefaultCNI:false CNI: NodeIP: NodePort:8443 NodeName:} Nodes:[{Name: IP:192.168.49.2 Port:8443 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}] Addons:map[] VerifyComponents:map[apiserver:true apps_running:true default_sa:true extra:true kubelet:true node_ready:true system_pods:true] StartHostTimeout:6m0s ScheduledStop:<nil> ExposedPorts:[] ListenAddress: Network: MultiNodeRequested:false}
```
extract:
> Nodes:[{Name: IP:192.168.49.2 Port:8443 KubernetesVersion:v1.20.2 ControlPlane:true Worker:true}]

so, i've replaced code segment in this pr:
```
		name := cfg.Name
		if n.Name != "" {
			name = fmt.Sprintf("%s-%s", cfg.Name, n.Name)
		}
```
with:
> name := config.MachineName

https://github.com/kubernetes/minikube/blob/81eeb7dd979e29a712ac5153fc4b6b40bd519f9f/pkg/minikube/config/profile.go#L328-L335

and that should fix it
also, i've rebased to upstream

// side thought: i'm not sure why none driver sets the name of its node (in contrast to other drivers)?

medyagh reviewed (approved):
thank you @prezha this PR made the Docker Linux Tests have 0 Failures !!! nice work
